shelley/chain-and-ledger/executable-spec/src/Cardano/Ledger/Chain.hs

```diff
@@ -0,0 +1,159 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Ledger.Chain
+  ( -- | Chain Checks
+    ChainChecksData (..),
+    ChainPredicateFailure (..),
+    pparamsToChainChecksData,
+    chainChecks,
+    -- | Ada Pots
+    AdaPots (..),
+    totalAdaES,
+    totalAdaPotsES,
+  )
+where
+
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Era (Crypto, Era)
+import Cardano.Ledger.Shelley.Constraints (UsesValue)
+import qualified Cardano.Ledger.Val as Val
+import Cardano.Protocol.TPraos.BHeader
+  ( BHeader,
+    bHeaderSize,
+    bhbody,
+    hBbsize,
+  )
+import Control.Monad (unless)
+import Control.Monad.Except (MonadError, throwError)
+import Data.Foldable (fold)
+import qualified Data.Map.Strict as Map
+import GHC.Generics (Generic)
+import GHC.Records
+import NoThunks.Class (NoThunks (..))
+import Numeric.Natural (Natural)
+import Shelley.Spec.Ledger.LedgerState
+  ( AccountState (..),
+    DPState (..),
+    DState (..),
+    EpochState (..),
+    LedgerState (..),
+    UTxOState (..),
+  )
+import Shelley.Spec.Ledger.PParams (ProtVer (..))
+import Shelley.Spec.Ledger.UTxO (balance)
+
+data ChainChecksData = ChainChecksData
+  { ccMaxBHSize :: Natural,
+    ccMaxBBSize :: Natural,
+    ccProtocolVersion :: ProtVer
+  }
+  deriving (Show, Eq, Generic, NoThunks)
+
+pparamsToChainChecksData ::
+  ( HasField "_maxBHSize" pp Natural,
+    HasField "_maxBBSize" pp Natural,
+    HasField "_protocolVersion" pp ProtVer
+  ) =>
+  pp ->
+  ChainChecksData
+pparamsToChainChecksData pp =
+  ChainChecksData
+    { ccMaxBHSize = getField @"_maxBHSize" pp,
+      ccMaxBBSize = getField @"_maxBBSize" pp,
+      ccProtocolVersion = getField @"_protocolVersion" pp
+    }
+
+data ChainPredicateFailure era
+  = HeaderSizeTooLargeCHAIN
+      !Natural -- Header Size
+      !Natural -- Max Header Size
+  | BlockSizeTooLargeCHAIN
+      !Natural -- Block Size
+      !Natural -- Max Block Size
+  | ObsoleteNodeCHAIN
+      !Natural -- protocol version used
+      !Natural -- max protocol version
+  deriving (Generic, Show, Eq, Ord)
+
+instance NoThunks (ChainPredicateFailure era)
+
+chainChecks ::
+  ( Era era,
+    MonadError (ChainPredicateFailure era) m
+  ) =>
+  Natural ->
+  ChainChecksData ->
+  BHeader (Crypto era) ->
+  m ()
+chainChecks maxpv ccd bh = do
+  unless (m <= maxpv) $ throwError (ObsoleteNodeCHAIN m maxpv)
+  unless (fromIntegral (bHeaderSize bh) <= ccMaxBHSize ccd) $
+    throwError $
+      HeaderSizeTooLargeCHAIN (fromIntegral $ bHeaderSize bh) (ccMaxBHSize ccd)
+  unless (hBbsize (bhbody bh) <= ccMaxBBSize ccd) $
+    throwError $
+      BlockSizeTooLargeCHAIN (hBbsize (bhbody bh)) (ccMaxBBSize ccd)
+  where
+    (ProtVer m _) = ccProtocolVersion ccd
+
+data AdaPots = AdaPots
```

Comment:
Do we actually use this anywhere? If so, I'd suggest that putting it into the wallet API module or similar might be a better place. At the moment I can only find its usage in tests.

Reply:
It's probably only in tests. I'll move it to the tests if so, and otherwise to same API module, thanks!

Reply:
db-sync uses it, I'll move it to the wallet API module